manpages/man/tiledb_array_define.1
.TH tiledb_array_define 1 "09 October 2015" "Version 0.1" "TileDB programs"
 
.SH NAME
tiledb_array_define - defines an array, specifying an array schema

.SH SYNOPSIS
.B tiledb_array_define 
.BI "-A " "array_name " "-a " "attribute_name_1,... " "-d " "dim_name_1,... " 
.br
.BI "-D " "dim_domain_1_low,dim_domain_1_high,... " "-t " "type_1,... " 
.br
[
.BI "-w " "workspace "
] [
.BI "-g " "group "
] [
.BI "-o " "cell_order"
] [
.BI "-s " "consolidation_step"
] 
.br
[ [ [
.BI "-e " "tile_extent_1,..."
] [
.BI "-O " "tile_order"
] ] | 
[ 
.BI "-c " "capacity"
] ] 
.br
[ 
.BI "-z " "compression"
] 

.SH DESCRIPTION
.B tiledb_array_define
defines an array, specifying its schema. Every array must be defined prior to
its use. On error, it returns \fB-1\fR. Moreover, if it is compiled in 
\fBVERBOSE\fR mode, it prints comprehensive error messages on \fIstderr\fR. 
On success, it returns \fB0\fR.
The following conventions apply to the \fBSYNOPSIS\fR section:

.TS
tab (@);
c lx .
\fBbold text\fR @ type exactly as shown
\fIitalic text\fR @ replace with appropriate argument
[\fB\-a \fIarg\fR]@ any or all options within [ ] are optional
\fB\-a \fIarg\fR | \fB\-b \fIarg\fR @ \
options delimited by | cannot be used together
.TE

Each array is comprised of a set of \fBdimensions\fR, and a set of 
\fBattributes\fR. Each array \fBcell\fR is essentially a tuple, consisting of a
set of dimension values, collectively called as \fIcoordinates\fR, and a set of
\fIattribute\fR values. The coordinates and attribute values may be of different
\fBtypes\fR, all coordinates though must have the same type. The 
coordinates draw their values from their corresponding \fBdimension domains\fR.
A cell may have a \fIfixed\fR or a \fIvariable\fR number of values on each
attribute. TileDB stores internally only the \fInon-empty\fR (i.e., non-null,
non-zero) cells. 

An array is internally represented as a set of \fBfragments\fR (i.e., snapshots
of the array corresponding to a subset of cells), each consisting of a set of 
\fBtiles\fR. Each tile is essentially a \fIhyper-rectangle\fR in the
\fIlogical\fR (i.e., dimension) space that groups a set of non-empty cells. The
tiles of an array may be \fIregular\fR or \fIirregular\fR. Regular tiles have
\fIfixed\fR \fBtile extents\fR across each dimension, i.e., they have the same
shape in the logical space, but may have a different non-empty cell
\fBcapacity\fR. Irregular tiles have a \fIfixed\fR non-empty cell capacity, but
may have \fIvariable shape\fR in the logical space.

Each array stores its cells internally in a \fIsorted\fR \fBtile order\fR and 
\fBcell order\fR. For the case of irregular tiles, the tile order is implied
by the cell order. We refer to the sorted order of the cells as the 
\fBglobal cell order\fR. 

The user may define a \fBcompression\fR type \fIper attribute\fR. The
compression is performed on a tile basis (implying also per fragment, per 
attribute).

Finally, TileDB updates arrays in \fIbatches\fR, i.e., it modifies 
\fIsets of cells\fR instead of single cells. The updates can be insertions
of new cells, deletions of old ones, or simple modifications. All updates
though are treated essentially as insertions. When a new
set of cells is inserted into an array, TileDB initially creates a new
array \fBfragment\fR encompassing only the updates. Periodically, as new
updates arrive, TileDB \fBconsolidates\fR multiple fragments into a single one.
The consolidation frequency is defined by a \fBconsolidation step\fR parameter.
If this parameter is 1, every new fragment is always consolidated with the
existing one. If it is larger than 1, then consolidation occurs in a 
\fIhierarchical\fR manner. For instance, if \fBs\fR is the consolidation step,
each new fragment essentially represents a leaf of a complete \fBs\fR-ary 
\fBmerge tree\fR that is constructed bottom-up. When \fBs\fR fragment-leaves are
created, they are consolidated into a single one, which becomes their parent 
node in the tree, and the leaves are disregarded for the rest of the system
lifetime. In general, whenever \fBs\fR fragment nodes are created at the
same level, they are consolidated into a new fragment that becomes
their parent, and these nodes are disregarded thereafter.

All the data of an array are included in a directory in the underlying file
system, which has the same name as the array. Each array has a file called 
\fIarray_schema\fR under its directory, which stores its schema information.
Multiple arrays can be organized into \fBgroups\fR, where a group is a directory
containing array directories and/or other group directories. Each group
directory contains an empty file called \fB.tiledb_group\fR, which is used
for sanity checks. A \fBworkspace\fR can contain groups and/or arrays, whereas
it can also be a group itself. Groups enable the hierarchical organization of
arrays, which follows naturally the directory organization in file systems. The
workspace practically constitutes the root of this organization. Each workspace
directory contains an empty file called \fB.tiledb_workspace\fR, which is used
for sanity checks. 
 
.SH OPTIONS
.TP
.BI "-w" " workspace" "\fR, " \fB --workspace=\fIworkspace\fR  
The path to the workspace folder where TileDB arrays can be stored. Note that
all non-existent directories in the workspace path will be created. Moreover,
the caller must have read and write permissions on the workspace directory.
If it is not provided, the current working directory is set as the workspace
by default. \fINote:\fR The workspace name must be in POSIX format, i.e., 
it must consist only of alphanumerics and characters '_', '\-' and '.',
otherwise the result is unpredictable.

.TP
.BI "-g" " group" "\fR, " \fB --group=\fIgroup\fR  
This is a directory inside the workspace where the array data will be stored.
Note that any group path provided is converted into an absolute path 
\fBrelative to the workspace directory\fR (examples are provided below), i.e., 
regarding all home ("\\~/"), current ("./") and root ("/") as the workspace
directory. The groups allow the hierarchical organization of TileDB arrays into
sub-directories within the workspace directory. If the group is not provided, 
the workspace is set as the default group. \fINote:\fR The workspace name
must be in POSIX format, i.e., it must consist only of alphanumerics and 
characters '_', '\-' and '.', otherwise the result is unpredictable.

.TP
.BI "-A" " array_name" "\fR, " \fB --array-name=\fIarray_name\fR  
The name of the array whose schema is being defined. \fINote:\fR Currently, 
TileDB supports only \fBPOSIX\fR array names, i.e., an array name can contain 
only alphanumerics and characters '_', '\-' and '.'. 

.TP
.BI "-a" " attribute_name_1,..." "\fR, " \
\fB --attribute-names=\fIattribute_name_1,...\fR  
The names of the attributes of the array. Similarly to the array name, they 
should be in \fBPOSIX\fR format.

.TP
.BI "-d" " dim_name_1,..." "\fR, " \
\fB --dim-names=\fIdim_name_1,...\fR  
The names of the dimensions of the array. Similarly to the array name, they 
should be in \fBPOSIX\fR format. A dimension cannot have the same name as
an attribute. 

.TP
.BI "-D" " dim_domain_1_low,dim_domain_1_high,..." "\fR, " \
\fB --dim-domains=\fIdim_domain_1_low,dim_domain_1_high,...\fR  
The domains of the dimensions. There should be a [low,high] pair for
every dimension, whose order must follow that of the dimension names
in the \fB\-d\fR option. 

.TP
.BI "-t" " type_1,..." "\fR, " \
\fB --types=\fItype_1,...\fR  
The types of the attributes and (collectively) of the coordinates. Specifically,
if there are \fIattribute_num\fR attributes, there should be provided 
\fIattribute_num+1\fR types. The types of the attributes must be given
first, following the order of the attribute names in the \fB\-a\fR option.
The type of the coordinates must be appear last. The supported attribute
types are \fBchar\fR, \fBint32\fR, \fBint64\fR, \fBfloat32\fR, and
\fBfloat64\fR. The supported coordinate types are \fBint32\fR, \fBint64\fR, 
\fBfloat32\fR, and \fBfloat64\fR. 

Optionally, one may specify the number of values to be stored for a particular
attribute in each cell. This is done by appending '\fB:\fR' followed by the 
desired number of values after the type (e.g., \fBint32:3\fR). If no such value
is provided, the default is 1. If one needs to specify that an attribute may
take a variable (unknown a priori) number of values, '\fB:var\fR' must
be appended after the type (e.g., \fBchar:var\fR). Note that the dimension
type cannot have multiple values; a single set of coordinates uniquely 
identifies each cell.

.TP 
.BI "-e" " tile_extent_1,..." "\fR, " \
\fB --tile_extents=\fItile_extent_1,...\fR  
It specifies the extent of a tile across each dimension. If this option is
included, then the array will have \fIregular\fR tiles; if it is omitted, the
array will have \fIirregular\fR tiles. If there are \fIdim_num\fR dimensions,
there should be provided \fIdim_num\fR tile extents, following the same order
as that of the dimension names in the \fB-d\fR option. Each tile extent must
be a non-negative real number that does not exceed the corresponding domain 
size.

.TP 
.BI "-o" " cell_order"  "\fR, " \
\fB --cell_order=\fIcell_order\fR  
The order in which the cells will be stored internally. The supported orders
are \fBrow-major\fR, \fBcolumn-major\fR and \fBhilbert\fR (i.e., following
the Hilbert space-filling curve). If no cell order is provided, the default
is \fBrow-major\fR.  

.TP 
.BI "-O" " tile_order"  "\fR, " \
\fB --tile_order=\fItile_order\fR  
The order in which the tiles will be stored internally. The supported orders
are \fBrow-major\fR, \fBcolumn-major\fR and \fBhilbert\fR (i.e., following
the Hilbert space-filling curve). If no cell order is provided, the default
is \fBrow-major\fR. This is applicable only to regular tiles.

.TP 
.BI "-c" " capacity"  "\fR, " \
\fB --capacity=\fIcapacity\fR  
This specifies the fixed number of non-empty cells stored in each tile. It
is applicable only to irregular tiles and, hence, cannot be used together
with options \fB\-e\fR and \fB\-O\fR. If it is not provided, a default
value is used. 

.TP 
.BI "-s" " consolidation_step"  "\fR, " \
\fB --consolidation_step=\fIconsolidation_step\fR  
It specifies the frequency of fragment consolidation, described in section 
\fBDESCRIPTION\fR. If it is not provided, the default is 1.

.TP 
.BI "-z" " compression_type"  "\fR, " \
\fB --compression=\fIcompression_type\fR  
It specifies the per-attribute compression type. Currently, there are two
types supported: \fBNONE\fR and \fBGZIP\fR. The user must input as many
types as the number of attributes (in the order of appearence in the 
\fB\-a\fR option), plus one in the end for the coordinates. If omitted,
the default is \fBNONE\fR for all the attributes and coordinates.

.SH EXAMPLES
.TP
\fBtiledb_array_define\fR \\ 
    \fB-A \fRmy_array \\
    \fB-w \fRmy_workspace/ \\
    \fB-g \fR\\~/my_group/ \\
    \fB-a \fRattr1,attr2,attr3 \\
    \fB-d \fRdim1,dim2 \\
    \fB-D \fR0,100,0,200 \\
    \fB-t \fRint:3,float64,char:var,int64 \\
    \fB-o \fRhilbert \\
    \fB-c \fR1000 \\
    \fB-s \fR5

This defines array \fImy_array\fR, whose data are stored in directory
\fI./my_workspace/my_group/my_array\fR. Observe how the group path is translated
with respect to the workspace directory. The array has \fI3\fR attributes and
\fI2\fR dimensions. Dimension \fIdim1\fR has domain \fI[0,100]\fR, \fIdim2\fR
has domain \fI[0,200]\fR. The coordinates are of type \fBint64\fR. Attribute
\fIattr1\fR is of type \fBint\fR, and each cell always stores \fI3\fR values on
this attribute. Attribute \fIattr2\fR is of type \fBfloat64\fR, and each cell
always stores \fI1\fR value on this attribute. Attribute \fIattr3\fR is of type
\fBchar\fR, and each cell stores a \fIvariable\fR number of values on this 
attribute (i.e., it stores arbitrary character strings). The array has 
\fIirregular\fR tiles. The cell order is \fBhilbert\fR. Each tile accommodates 
exactly \fI1000\fR (non-empty) cells. The consolidation step is set to \fI5\fR. 
There will be no compression.

.TP
\fBtiledb_array_define\fR \\ 
    \fB-A \fRmy_array \\
    \fB-a \fRattr1,attr2,attr3 \\
    \fB-d \fRdim1,dim2 \\
    \fB-D \fR0,100,0,200 \\
    \fB-e \fR10,20 \\
    \fB-t \fRint:3,float64,char:var,int64 \\
    \fB-O \fRrow-major \\
    \fB-o \fRhilbert \\
    \fB-s \fR5 \\
    \fB-z \fRGZIP,GZIP,NONE,NONE

This is similar to the previous example, but now the array has \fIregular\fR 
tiles, and it is stored in directory \fI./my_array\fR since no workspace and
group are provided (and, hence, the current working directory is used as both
the workspace and group by default). In detail, it defines 
array \fImy_array\fR in workspace folder \fImy_workspace\fR. The array has 
\fI3\fR attributes and \fI2\fR dimensions. Dimension \fIdim1\fR has domain 
\fI[0,100]\fR, \fIdim2\fR has domain \fI[0,200]\fR. The coordinates are of type 
\fBint64\fR. Attribute \fIattr1\fR is of type \fBint\fR, and each cell always 
stores \fI3\fR values on this attribute. Attribute \fIattr2\fR is of type 
\fBfloat64\fR, and each cell always stores \fI1\fR value on this attribute. 
Attribute \fIattr3\fR is of type \fBchar\fR, and each cell stores a 
\fIvariable\fR number of values on this attribute (i.e., it stores arbitrary 
character strings). The array has \fIregular\fR tiles. Each tile has (logical)
size \fI10x20\fR. The tile order is \fBrow-major\fR, whereas the cell order is 
\fBhilbert\fR. The consolidation step is set to \fI5\fR. GZIP ompression will be
used for attributes \fIattr1\fR and \fIattr2\fR, and no compression for 
\fIattr3\fR and the coordinates. 

.SH EXIT STATUS
.TP 
\fB0\fR for success and \fB-1\fR for error.

.SH SEE ALSO
.BR tiledb_array_schema_show (1)

.SH AUTHOR
Stavros Papadopoulos <http://people.csail.mit.edu/stavrosp>, Copyright (c) 2015


tiledb_cmd/src/tiledb_array_define.cc
/**
 * @file   tiledb_array_define.cc
 * @author Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * @section LICENSE
 *
 * The MIT License
 * 
 * @copyright Copyright (c) 2015 Stavros Papadopoulos <stavrosp@csail.mit.edu>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * 
 * @section DESCRIPTION
 *
 * Implements command "tiledb_array_define", which enables the user to define
 * an array, specifying its array schema.
 */

#include "csv_line.h"
#include "special_values.h"
#include "tiledb.h"
#include <getopt.h>
#include <iostream>
#include <string>

#define PRINT_ERROR(x) std::cerr << "[TileDB] Error: " << x << ".\n"
#define PRINT(x) std::cout << "[TileDB] " << x << ".\n"

/* 
 * Parses the command options. It returns the array workspace, group, and 
 * array schema. The array schema is serialized in CSV string array_schema_str.
 * The format of this string is the following (single line, no '\n' characters
 * involved):
 * 
 * array_name,attribute_num,attribute_1,...,attribute_{attribute_num},
 * dim_num,dim_1,...,dim_{dim_num},
 * dim_domain_1_low,dim_domain_1_high,...,
 * dim_domain_{dim_num}_low,dim_domain_{dim_num}_high
 * type_1,...,type_{attribute_num+1}
 * tile_extents_1,...,tile_extents_{dim_num},
 * cell_order,tile_order,capacity,consolidation_step,
 * compression_type_1,...,compression_type_{attribute_num+1}
 *
 * If one of the items is omitted (e.g., tile_order), then this CSV field will
 * contain character '*' (e.g., it should be ...,cell_order,*,capacity,...). 
 * 
 * It returns 0 on success. On error, it prints a message on stderr if compiled
 * in VERBOSE mode and returns -1.
 */                                                            
int parse_options(
    int argc, 
    char** argv, 
    std::string& workspace,
    std::string& group,
    std::string& array_schema_str) {

  // Initialization
  workspace = "";
  group = "";
  std::string array_name("");
  std::string attribute_names_str("");
  std::string dim_names_str("");
  std::string dim_domains_str("");
  std::string types_str("");
  std::string tile_extents_str("");
  std::string cell_order_str("");
  std::string tile_order_str("");
  std::string capacity_str("");
  std::string consolidation_step_str("");
  std::string compression_str("");

  // Auxiliary variables
  CSVLine temp, array_schema_csv;

  // *************** //
  // Parse options * //
  // *************** // 

  // Define long options
  static struct option long_options[] = 
  {
    {"attribute-names",1,0,'a'},
    {"array-name",1,0,'A'},
    {"capacity",1,0,'c'},
    {"dim-names",1,0,'d'},
    {"dim-domains",1,0,'D'},
    {"tile-extents",1,0,'e'},
    {"group",1,0,'g'},
    {"cell-order",1,0,'o'},
    {"tile-order",1,0,'O'},
    {"consolidation-step",1,0,'s'},
    {"types",1,0,'t'},
    {"workspace",1,0,'w'},
    {"compression",1,0,'z'},
    {0,0,0,0},
  };

  // Define short options
  const char* short_options = "a:A:c:d:D:e:g:o:O:s:t:w:z:";
  // Get options
  int c;
  int option_num = 0;
  while((c=getopt_long(argc, argv, short_options, long_options, NULL)) >= 0) {
    ++option_num;
    switch(c) {
      case 'a':
        if(attribute_names_str != "") {
          PRINT_ERROR("More than one attribute name lists provided");
          return -1;
        }
        attribute_names_str = optarg;
        break;
      case 'A':
        if(array_name != "") {
          PRINT_ERROR("More than one array names provided");
          return -1;
        }
        array_name = optarg;
        break;
      case 'c':
        if(capacity_str != "") {
          PRINT_ERROR("More than one capacities provided");
          return -1;
        }
        capacity_str = optarg;
        break;
      case 'd':
        if(dim_names_str != "") {
          PRINT_ERROR("More than one dimension name lists provided");
          return -1;
        }
        dim_names_str = optarg;
        break;
      case 'D':
        if(dim_domains_str != "") {
          PRINT_ERROR("More than one dimension domain lists provided");
          return -1;
        }
        dim_domains_str = optarg;
        break;
      case 'e':
        if(tile_extents_str != "") {
          PRINT_ERROR("More than one tile extent lists provided");
          return -1;
        }
        tile_extents_str = optarg;
        break;
      case 'g':
        if(group != "") {
          PRINT_ERROR("More than one groups provided");
          return -1;
        }
        group = optarg;
        break;
      case 'o':
        if(cell_order_str != "") {
          PRINT_ERROR("More than one cell orders provided");
          return -1;
        }
        cell_order_str = optarg;
        break;
      case 'O':
        if(tile_order_str != "") {
          PRINT_ERROR("More than one tile orders provided");
          return -1;
        }
        tile_order_str = optarg;
        break;
      case 's':
        if(consolidation_step_str != "") {
          PRINT_ERROR("More than one consolidation steps provided");
          return -1;
        }
        consolidation_step_str = optarg;
        break;
      case 't':
        if(types_str != "") {
          PRINT_ERROR("More than one type lists provided");
          return -1;
        }
        types_str = optarg;
        break;
      case 'w':
        if(workspace != "") {
          PRINT_ERROR("More than one workspaces provided");
          return -1;
        }
        workspace = optarg;
        break;
      case 'z':
        if(compression_str != "") {
          PRINT_ERROR("More than one compression type lists provided");
          return -1;
        }
        compression_str = optarg;
        break;
      default:
        return -1;
    }
  }

  // ******************* //
  // Check correctness * //
  // ******************* //

  // ----- Check number of arguments -----
  if((argc-1) != 2*option_num) {
    PRINT_ERROR("Arguments-options mismatch"); 
    return -1;
  }
  // ----- Check if correct arguments have been given ----- //
  if(array_name == "") {
    PRINT_ERROR("Array name not provided");
    return -1;
  }
  if(attribute_names_str == "") {
    PRINT_ERROR("Attribute names not provided");
    return -1;
  }
  if(dim_names_str == "") {
    PRINT_ERROR("Dimension names not provided");
    return -1;
  }
  if(dim_domains_str == "") {
    PRINT_ERROR("Dimension domains not provided");
    return -1;
  }
  if(types_str == "") {
    PRINT_ERROR("Types not provided");
    return -1;
  }
  // ----- The capacity cannot be set for regular tiles ----- //
  if(tile_extents_str != "" && capacity_str != "") {
    PRINT_ERROR("Capacity is meaningless in the case of regular tiles");
    return -1;
  }
  // ----- Tile order must not be given if the tiles are irregular ----- //
  if(tile_extents_str == "" && tile_order_str != "") {
    PRINT_ERROR("Tile order is meaningless in the case of irregular tiles");
    return -1;
  } 
  // ----- Tile order must not be given together with capacity ----- //
  if(capacity_str != "" && tile_order_str != "") {
    PRINT_ERROR("It is meaningless to provide both tile order and capacity");
    return -1;
  } 

  // ----- Check workspace and group multiplicities ----- //
  temp.clear();
  temp << workspace;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one workspaces provided");
    return -1;
  }
  temp.clear();
  temp << group;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one groups provided");
    return -1;
  }

  // ********************************** //
  // Serialize into a single CSV string //
  // ********************************** //

  // ----- array name ----- //
  temp.clear();
  temp << array_name;
  if(temp.val_num() > 1) {
    PRINT_ERROR("More than one array names provided");
    return -1;
  }
  array_schema_csv << temp;

  // ----- attribute_num and attribute names ----- //
  temp.clear();
  temp << attribute_names_str;
  int attribute_num = temp.val_num();
  array_schema_csv << attribute_num;
  array_schema_csv << temp;

  // ----- dim_num and dimension names ----- //
  temp.clear();
  temp << dim_names_str;
  int dim_num = temp.val_num();
  array_schema_csv << dim_num;
  array_schema_csv << temp;

  // ----- dimension domains ----- //
  temp.clear();
  temp << dim_domains_str;
  if(temp.val_num() != 2*dim_num) {
    PRINT_ERROR("The number of domain bounds does not match the "
                "provided number of dimensions");
    return -1;
  }
  array_schema_csv << temp;

  // ----- types ----- // 
  temp.clear();
  temp << types_str;
  if(temp.val_num() != attribute_num + 1) {
    PRINT_ERROR("The number of types does not match the number of "
                "attributes");
    return -1;
  }
  array_schema_csv << temp;

  // ----- tile extents ----- //
  temp.clear();
  if(tile_extents_str == "") {
    array_schema_csv << NULL_CHAR;
  } else {
    temp << tile_extents_str;
    if(temp.val_num() != dim_num) {
      PRINT_ERROR("The number of tile extents does not match the number of "
                  "dimensions");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ----- cell order ----- //
  temp.clear();
  if(cell_order_str == "") {
    array_schema_csv << NULL_CHAR;
  } else {
    temp << cell_order_str;
    if(temp.val_num() > 1) {
      PRINT_ERROR("More than one cell orders provided");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ----- tile order ----- //
  temp.clear();
  if(tile_order_str == "") {
    array_schema_csv << NULL_CHAR;
  } else {
    temp << tile_order_str;
    if(temp.val_num() > 1) {
      PRINT_ERROR("More than one tile orders provided");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ----- capacity ----- //
  temp.clear();
  if(capacity_str == "") {
    array_schema_csv << NULL_CHAR;
  } else {
    temp << capacity_str;
    if(temp.val_num() > 1) {
      PRINT_ERROR("More than one capacities provided");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ----- consolidation step ----- //
  temp.clear();
  if(consolidation_step_str == "") {
    array_schema_csv << NULL_CHAR;
  } else {
    temp << consolidation_step_str;
    if(temp.val_num() > 1) {
      PRINT_ERROR("More than one consolidation steps provided");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ----- compression ----- // 
  temp.clear();
  temp << compression_str;
  if(compression_str == "") {
    array_schema_csv << NULL_CHAR; 
  } else {
    if(temp.val_num() != attribute_num + 1) {
      PRINT_ERROR("The number of compression types does not match the number "
                  "of attributes");
      return -1;
    }
    array_schema_csv << temp;
  }

  // ********************************* //
  // Set the array schema (CSV) string //
  // ********************************* //
  array_schema_str = array_schema_csv.c_str();

  // Success
  return 0;
}

int main(int argc, char** argv) {
  // Parse command line
  std::string workspace, group, array_schema_str;
  if(parse_options(argc, argv, workspace, group, array_schema_str)) { 
    PRINT_ERROR("Program failed");
    return -1;
  }

  // Initialize TileDB
  TileDB_CTX* tiledb_ctx;
  if(tiledb_ctx_init(tiledb_ctx))
    return -1;

  // Store the array schema
  if(tiledb_array_define(tiledb_ctx, workspace.c_str(), group.c_str(), 
                         array_schema_str.c_str())) {
    tiledb_ctx_finalize(tiledb_ctx);
    PRINT_ERROR("Program failed");
    return -1;
  }

  // Finalize TileDB
  if(tiledb_ctx_finalize(tiledb_ctx))
    return -1;

  // Success
  PRINT("Program executed successfully");
  
  return 0;
}
